text/0000-export.md

```diff
@@ -0,0 +1,398 @@
+
+For example, an exported item in `foo::bar` in the crate `mycrate` would be exported with a symbol name such as:
+
+```
+_RNvNtC_7mycrate3foo3bar_f8771d213159376fafbff1d3b93bb212
+```
+
+Where the first part is the (mangled) path and name of the item,
+and the second part is the hexadecimal representation of a 128 hash of all relevant signature and type information.
+The hash algorithm is still to be determined.
+
+(See also the "alternatives" section below.)
+
+### Type Information
+
+As mentioned above, the hash in a symbol name needs to cover _all_ relevant type information.
+However, exactly which information is and isn't relevant for safety is a complicated question.
+
+#### Types with Public Fields
+
+For a simple user defined type where all fields are both public, like the `Point` struct below,
+the relevant parts are the size, alignment, and recursively all field information.
+
+```rust
+#[export]
+#[repr(C)]
+pub struct Point {
+    pub x: f32,
+    pub y: f32,
+    pub name: &str,
+}
+```
+
+The `#[export]` attribute is the user's commitment to keep the type stable, but without `unsafe`,
+any mistakes should _not_ result in unsoundness.
+Accidentally changing the struct to swap the `x` and `name` fields should result in a different hash,
+such that the `f32` won't get interpreted as a `&str`, for example.
+
+Note that, technically, the names of the type and the fields are not relevant, at least _not for memory safety_.
+Swapping the `x` and `y` fields result in surprises and bugs and shouldn't be done,
+but it won't result in undefined behaviour, since any Rust code can swap the fields without using `unsafe`.
+
+However, for public fields, the field names are already part of the stable API, so we include them in the hash as well.
+
+It is an error to use a plain `#[export]` attribute on a type with out stable `#[repr(…)]`,
+if it has any private fields,
+or if any of the fields are not of an `#[export]`ed or builtin type.
+
+#### Types with Private Fields
+
+For types where not all fields are public, the situation is much more complicated.
+
+Private fields usually come with certain *invariants*, and come with `unsafe` code that makes assumptions about them.
+For example, the private fields of a `Vec` are assumed to represent a valid "owned" pointer to an allocation together with its capacity and initialized size.
+
+If it would be possible to define a identically named type with the same fields but different (or no) invariants/assumptions,
+or just change the invariants in an existing library,
+it'd be possible to cause undefined behavior by loading the "wrong" dynamic library.
+
+Therefore, we can't allow a regular `#[export]` attribute on a type with private fields,
+since we have no way of automatically determining the invariants / unsafe assumptions about private fields.
+
+Instead, for these types, we must require the user to *unsafely* commit to
```

Comment:
One thing I didn't see mentioned here is whether the *size* of the type is considered. If a type ADDS a private field, this could (potentially) invalidate assumptions when passing by value.

For example if a type starts as `struct A { pub x: i32, y: u32 }`, and someone marks it as unsafe stable, can we help them at all if they make it `struct A { pub x: i32 }` or `struct A { pub x: i32, y: u32, z: u8 }`?

Reply:
I've clarified the RFC to mention this.

Reply:
Got it! I agree that having a way of forcing "by reference only" is good! I was more focused on "they've opted-in to unsafe snapshotting", but they *do* change the size, which is always going to cause UB on use. I'm unsure if it is still possible to verify the new size and the claimed size from only the hash, it just might be worth also explicitly stating the size or requiring export opaque to help avoid silent corruptions.

e.g. requiring one of:

```rust
// maybe size could also be `{ 2 * size_of::<usize>() }` or something else const for portability
#[export(unsafe_stable_abi = "ca83050b302bf0644a1417ac3fa6982a", size = 8)]
```

or

```
#[export(unsafe_stable_abi = "ca83050b302bf0644a1417ac3fa6982a", opaque)]
```

or having it possible to "decode" the size from the abi tag (e.g. part of the hash AND stored as plaintext somehow)

Reply:
Yeah maybe. I don't think we should do `size = 8`, but maybe we should require either `sized` or `opaque`. But I'm hoping to leave opaque types to a future proposal for now.